db/sqlglue.c

```diff
@@ -591,6 +591,49 @@ void sql_get_query_id(struct sql_thread *thd)
     }
 }
 
+static unsigned int query_path_component_hash(const void *key, int len)
+{
+    const struct query_path_component *q = key;
+    const char *name;
+    if (q->fdb == NULL) {
+        name = q->lcl_tbl_name;
+    } else {
+        name = fdb_table_entry_tblname(q->fdb);
+    }
+    struct {
+        int ix;
+        char name[strlen(name) + 1];
+    } x;
+    x.ix = q->ix;
+    strcpy(x.name, name);
+    return hash_default_fixedwidth((void*)&x, sizeof(x));
+}
+
+static int query_path_component_cmp(const void *key1, const void *key2, int len)
+{
+    const char *n1, *n2;
+    const struct query_path_component *q1 = key1, *q2 = key2;
+    if (q1->fdb == NULL && q2->fdb == NULL) {
+        // both local
+        n1 = q1->lcl_tbl_name;
+        n2 = q2->lcl_tbl_name;
+    } else if (q1->fdb == NULL || q2->fdb == NULL) {
+        // mismatch
+        return 1;
+    } else {
+        // both remote
+        n1 = fdb_table_entry_tblname(q1->fdb);
+        n2 = fdb_table_entry_tblname(q2->fdb);
+    }
+    // check name + ix
+    if (strcmp(n1, n2) == 0) {
+        if (q1->ix == q2->ix) {
```

Comment:
Combine two if-s into one? Also I suggest you compare ix first then string compare as comparing 2 integers is a lot cheaper than comparing 2 strings.